tensorboard/plugins/graph/graph_util.py

```diff
@@ -0,0 +1,48 @@
+# -*- coding: utf-8 -*-
+# Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Utilities for graph plugin."""
+from tensorboard.compat.proto import function_pb2
+
+
+def combine_graph_defs(to_proto, from_proto):
+  """Combines two GraphDefs by adding nodes from from_proto into to_proto.
+
+  Args:
+    to_proto: a destination GraphDef.
+    from_proto: a GraphDef to copy contents from.
+
+  Returns:
+    to_proto
+  """
+  if from_proto.version != to_proto.version:
+    raise ValueError('Cannot combine GraphDefs of different versions.')
+
+  for from_node in from_proto.node:
+    to_node = to_proto.node.add()
```

Comment:
Correct me if I’m wrong: from our discussions, I took away that the
purpose of this function is to construct a union of _disjoint_ graphs so
that they may be visualized in the UI. But this function makes no
attempt to ensure or validate uniqueness of names: the result of this
function may be a `GraphDef` that has multiple node entries with the
same name, even though no graphs emitted by TensorFlow have that
property.

Does it make sense for graphs to have multiple nodes with the same name,
and will our code support that? (Does none of our data manipulation or
frontend UI logic depend on keying things uniquely by name?)

If the answer is something like, “this function is only expected to be
called with `to_proto` and `from_proto` representing function graphs
from the same TensorFlow run, so we know that their key sets are
disjoint”. In that case, we must verify that this assumption holds, and
raise a `ValueError` if the key sets are not disjoint.

Reply:
Not sure if this directly answers your question but we need something that can either render tree of graphs or combine the graphs perfectly. This is the first very naive attempt and we intend to iterate on it when we face issues after the alpha.

Now more concrete answers and assumptions:
- Akshay tells me that although nothing enforces and guarantees the global uniqueness, he says we are making every op's name globally unique. Perhaps this can change in the future and we will have to devise much more sophisticated strategy in that case. I have not found anything (did not spend time reading the code) that corroborates Akshay so I decided not to write that down in comment or anything.
- Graphlib (I may be wrong) seems to take the last instance of the node if the names (used as ID) collides. To answer your general question and your question in comment below here,
```
GraphDefA {
  nodeA {}
  nodeB {
    input: "node_a"
  }
}
GraphDefB {
  nodeA {}
  nodeC {
    input: "node_a"
  }
  nodeB {
    input: "node_c"
  }
}
```
will be combined like this
```
GraphDefComb {
  nodeA {}
  nodeB {
    input: "node_a"
  }
  nodeA {}
  nodeC {
    input: "node_a"
  }
  nodeB {
    input: "node_c"
  }
}
```
and it will visually render
```
nodeA {}
nodeC {
  input: "node_a"
}
nodeB {
  input: "node_c"
}
a -> c -> b
```

> In that case, we must verify that this assumption holds, and
raise a `ValueError` if the key sets are not disjoint.

That indeed is my assumption but I am a bit scared of adding invariants just yet since I am operating on very weakly proved assumptions and iterate on this.

Reply:
> > In that case, we must verify that this assumption holds, and
> > raise a `ValueError` if the key sets are not disjoint.
> 
> That indeed is my assumption but I am a bit scared of adding
> invariants just yet since I am operating on very weakly proved
> assumptions and iterate on this.

Then imho this is the _perfect_ place to assert that invariant: we have
an assumption and we want it to hold but we’re not actually sure if it
does. If it doesn’t, we sure want to know about it! The alternative is
for the code to silently not just “do the wrong thing” but “answer a
question for which we don’t even understand want the right answer
_should_ be”, leaving us to debug this down the line with far less
information.

Reply:
To be clear, the intent is for the graphs to always be rendered as
disjoint, right?

Reply:
> To be clear, the intent is for the graphs to always be rendered as
disjoint, right?

If there are multiple of them, yes (unless there is a use case we need not consider)

Reply:
Added a basic invariance. I honestly dunno what is supposed to happen for Function Libraries so I just kept that without any check.